--- a/net/minecraft/entity/projectile/DamagingProjectileEntity.java
+++ b/net/minecraft/entity/projectile/DamagingProjectileEntity.java
@@ -22,6 +22,8 @@
    public double field_70232_b;
    public double field_70233_c;
    public double field_70230_d;
+   public float bukkitYield = 1; // CraftBukkit
+   public boolean isIncendiary = true; // CraftBukkit
 
    protected DamagingProjectileEntity(EntityType<? extends DamagingProjectileEntity> p_i50173_1_, World p_i50173_2_) {
       super(p_i50173_1_, p_i50173_2_);
@@ -40,6 +42,11 @@
 
    }
 
+   // CraftBukkit - start
+   public void setDirection(double x, double y, double z) {
+   }
+   // CraftBukkit - end
+
    public DamagingProjectileEntity(EntityType<? extends DamagingProjectileEntity> p_i50175_1_, LivingEntity p_i50175_2_, double p_i50175_3_, double p_i50175_5_, double p_i50175_7_, World p_i50175_9_) {
       this(p_i50175_1_, p_i50175_2_.func_226277_ct_(), p_i50175_2_.func_226278_cu_(), p_i50175_2_.func_226281_cx_(), p_i50175_3_, p_i50175_5_, p_i50175_7_, p_i50175_9_);
       this.func_212361_a(p_i50175_2_);
@@ -69,7 +76,7 @@
          }
 
          RayTraceResult raytraceresult = ProjectileHelper.func_234618_a_(this, this::func_230298_a_, RayTraceContext.BlockMode.COLLIDER);
-         if (raytraceresult.func_216346_c() != RayTraceResult.Type.MISS) {
+         if (raytraceresult.func_216346_c() != RayTraceResult.Type.MISS && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult)) {
             this.func_70227_a(raytraceresult);
          }
 
