--- a/net/minecraft/entity/projectile/FireworkRocketEntity.java
+++ b/net/minecraft/entity/projectile/FireworkRocketEntity.java
@@ -36,11 +36,11 @@
    _interface = IRendersAsItem.class
 )
 public class FireworkRocketEntity extends ProjectileEntity implements IRendersAsItem {
-   private static final DataParameter<ItemStack> field_184566_a = EntityDataManager.func_187226_a(FireworkRocketEntity.class, DataSerializers.field_187196_f);
+   public static final DataParameter<ItemStack> field_184566_a = EntityDataManager.func_187226_a(FireworkRocketEntity.class, DataSerializers.field_187196_f); // private->public CraftBukkit
    private static final DataParameter<OptionalInt> field_191512_b = EntityDataManager.func_187226_a(FireworkRocketEntity.class, DataSerializers.field_218814_r);
-   private static final DataParameter<Boolean> field_213895_d = EntityDataManager.func_187226_a(FireworkRocketEntity.class, DataSerializers.field_187198_h);
+   public static final DataParameter<Boolean> field_213895_d = EntityDataManager.func_187226_a(FireworkRocketEntity.class, DataSerializers.field_187198_h); // private->public CraftBukkit
    private int field_92056_a;
-   private int field_92055_b;
+   public int field_92055_b; // private->public CraftBukkit
    private LivingEntity field_191513_e;
 
    public FireworkRocketEntity(EntityType<? extends FireworkRocketEntity> p_i50164_1_, World p_i50164_2_) {
@@ -155,6 +155,13 @@
 
    }
 
+   @Override
+   protected void func_70227_a(RayTraceResult result) {
+      if (result.func_216346_c() == RayTraceResult.Type.MISS || !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, result)) {
+         super.func_70227_a(result);
+      }
+   }
+
    private void func_213893_k() {
       this.field_70170_p.func_72960_a(this, (byte)17);
       this.func_191510_k();
