--- a/net/minecraft/entity/projectile/ShulkerBulletEntity.java
+++ b/net/minecraft/entity/projectile/ShulkerBulletEntity.java
@@ -30,6 +30,7 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.entity.EntityPotionEffectEvent.Cause;
 
 public class ShulkerBulletEntity extends ProjectileEntity {
    private Entity field_184571_b;
@@ -65,12 +66,25 @@
       this.field_184571_b = p_i46772_3_;
       this.field_184573_c = Direction.UP;
       this.func_184569_a(p_i46772_4_);
+      projectileSource = (org.bukkit.entity.LivingEntity) p_i46772_2_.getBukkitEntity(); // CraftBukkit
    }
 
    public SoundCategory func_184176_by() {
       return SoundCategory.HOSTILE;
    }
 
+   // CraftBukkit start
+   public Entity getTarget() {
+      return this.field_184571_b;
+   }
+
+   public void setTarget(final Entity e) {
+      this.field_184571_b = e;
+      this.field_184573_c = Direction.UP;
+      this.func_184569_a(Direction.Axis.X);
+   }
+   // CraftBukkit end
+
    protected void func_213281_b(CompoundNBT p_213281_1_) {
       super.func_213281_b(p_213281_1_);
       if (this.field_184571_b != null) {
@@ -214,7 +228,7 @@
          }
 
          RayTraceResult raytraceresult = ProjectileHelper.func_234618_a_(this, this::func_230298_a_, RayTraceContext.BlockMode.COLLIDER);
-         if (raytraceresult.func_216346_c() != RayTraceResult.Type.MISS) {
+            if (raytraceresult.func_216346_c() != RayTraceResult.Type.MISS && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult)) {
             this.func_70227_a(raytraceresult);
          }
       }
@@ -274,7 +288,7 @@
       if (flag) {
          this.func_174815_a(livingentity, entity);
          if (entity instanceof LivingEntity) {
-            ((LivingEntity)entity).func_195064_c(new EffectInstance(Effects.field_188424_y, 200));
+            ((LivingEntity) entity).addPotionEffect(new EffectInstance(Effects.field_188424_y, 200), Cause.ATTACK); // CraftBukkit
          }
       }
 
