--- a/net/minecraft/entity/projectile/ShulkerBulletEntity.java
+++ b/net/minecraft/entity/projectile/ShulkerBulletEntity.java
@@ -70,6 +70,18 @@
       return SoundCategory.HOSTILE;
    }
 
+   // CraftBukkit start
+   public Entity getTarget() {
+      return this.field_184571_b;
+   }
+
+   public void setTarget(final Entity e) {
+      this.field_184571_b = e;
+      this.field_184573_c = Direction.UP;
+      this.func_184569_a(Direction.Axis.X);
+   }
+   // CraftBukkit end
+
    protected void func_213281_b(CompoundNBT p_213281_1_) {
       super.func_213281_b(p_213281_1_);
       if (this.field_184571_b != null) {
@@ -213,7 +225,7 @@
          }
 
          RayTraceResult raytraceresult = ProjectileHelper.func_234618_a_(this, this::func_230298_a_);
-         if (raytraceresult.func_216346_c() != RayTraceResult.Type.MISS) {
+            if (raytraceresult.func_216346_c() != RayTraceResult.Type.MISS && !net.minecraftforge.event.ForgeEventFactory.onProjectileImpact(this, raytraceresult)) {
             this.func_70227_a(raytraceresult);
          }
       }
