--- a/net/minecraft/world/IBlockReader.java
+++ b/net/minecraft/world/IBlockReader.java
@@ -13,8 +13,10 @@
 import net.minecraft.util.math.BlockRayTraceResult;
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.util.math.RayTraceContext;
+import net.minecraft.util.math.RayTraceResult;
 import net.minecraft.util.math.shapes.VoxelShape;
 import net.minecraft.util.math.vector.Vector3d;
+import org.apache.commons.lang.NotImplementedException;
 
 public interface IBlockReader {
    @Nullable
@@ -25,7 +27,7 @@
    FluidState func_204610_c(BlockPos p_204610_1_);
 
    default int func_217298_h(BlockPos p_217298_1_) {
-      return this.func_180495_p(p_217298_1_).func_185906_d();
+      return this.func_180495_p(p_217298_1_).getLightValue(this, p_217298_1_);
    }
 
    default int func_201572_C() {
@@ -132,4 +134,8 @@
          }
       }
    }
+
+    default RayTraceResult rayTraceBlock(RayTraceContext rayTraceContext, BlockPos position) {
+      throw new NotImplementedException();
+    }
 }
