--- a/net/minecraft/world/Teleporter.java
+++ b/net/minecraft/world/Teleporter.java
@@ -10,8 +10,10 @@
 import net.minecraft.block.Blocks;
 import net.minecraft.block.NetherPortalBlock;
 import net.minecraft.block.pattern.BlockPattern;
+import net.minecraft.block.pattern.BlockPattern.PortalInfo;
 import net.minecraft.entity.Entity;
 import net.minecraft.entity.player.PlayerEntity;
+import net.minecraft.entity.player.ServerPlayerEntity;
 import net.minecraft.util.Direction;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.ChunkPos;
@@ -22,8 +24,9 @@
 import net.minecraft.village.PointOfInterestType;
 import net.minecraft.world.server.ServerWorld;
 import net.minecraft.world.server.TicketType;
+import org.apache.commons.lang.NotImplementedException;
 
-public class Teleporter {
+public class Teleporter implements net.minecraftforge.common.util.ITeleporter {
    protected final ServerWorld field_85192_a;
    protected final Random field_77187_a;
 
@@ -225,4 +228,17 @@
 
       return true;
    }
+
+   // CraftBukkit
+   public PortalInfo placeInExistingPortal(BlockPos blockpos, Vector3d vector3d, Direction teleportDirection, double x, double y, boolean b, int searchRadius) {
+      throw new NotImplementedException("todo: Magma come back");
+   }
+
+   public PortalInfo placeInPortal(ServerPlayerEntity serverPlayerEntity, BlockPos exitPosition, float f2, int searchRadius, boolean b) {
+      throw new NotImplementedException("todo: Magma come back");
+   }
+
+   public boolean makePortal(ServerPlayerEntity serverPlayerEntity, BlockPos exitPosition, int creationRadius) {
+      return false;
+   }
 }
