--- a/net/minecraft/block/AbstractFireBlock.java
+++ b/net/minecraft/block/AbstractFireBlock.java
@@ -12,6 +12,7 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.shapes.ISelectionContext;
 import net.minecraft.util.math.shapes.VoxelShape;
+import net.minecraft.world.DimensionType;
 import net.minecraft.world.IBlockReader;
 import net.minecraft.world.IWorld;
 import net.minecraft.world.World;
@@ -116,7 +117,13 @@
       if (!p_196262_4_.func_230279_az_()) {
          p_196262_4_.func_241209_g_(p_196262_4_.func_223314_ad() + 1);
          if (p_196262_4_.func_223314_ad() == 0) {
-            p_196262_4_.func_70015_d(8);
+            // CraftBukkit start
+            org.bukkit.event.entity.EntityCombustEvent event = new org.bukkit.event.entity.EntityCombustByBlockEvent(org.bukkit.craftbukkit.block.CraftBlock.at(p_196262_2_, p_196262_3_), p_196262_4_.getBukkitEntity(), 8);
+            p_196262_2_.getServerCB().getPluginManager().callEvent(event);
+            if (!event.isCancelled()) {
+               p_196262_4_.setFire(event.getDuration(), false);
+            }
+            // CraftBukkit end
          }
 
          p_196262_4_.func_70097_a(DamageSource.field_76372_a, this.field_235325_g_);
@@ -127,7 +134,8 @@
 
    public void func_220082_b(BlockState p_220082_1_, World p_220082_2_, BlockPos p_220082_3_, BlockState p_220082_4_, boolean p_220082_5_) {
       if (!p_220082_4_.func_203425_a(p_220082_1_.func_177230_c())) {
-         if (p_220082_2_.func_234923_W_() != World.field_234918_g_ && p_220082_2_.func_234923_W_() != World.field_234919_h_ || !NetherPortalBlock.func_176548_d(p_220082_2_, p_220082_3_)) {
+         // CraftBukkit - func_234922_V_()
+         if (p_220082_2_.func_234922_V_() != DimensionType.field_235999_c_ && p_220082_2_.func_234922_V_() != DimensionType.field_235999_c_ || !NetherPortalBlock.func_176548_d(p_220082_2_, p_220082_3_)) {
             if (!p_220082_1_.func_196955_c(p_220082_2_, p_220082_3_)) {
                p_220082_2_.func_217377_a(p_220082_3_, false);
             }
@@ -158,4 +166,13 @@
 
       return false;
    }
+
+   // CraftBukkit start
+   protected void fireExtinguished(IWorld world, BlockPos position) {
+      if (!org.bukkit.craftbukkit.event.CraftEventFactory.callBlockFadeEvent(world, position, Blocks.field_150350_a.func_176223_P()).isCancelled()) {
+         world.func_217377_a(position, false);
+      }
+   }
+   // CraftBukkit end
+
 }
