--- a/net/minecraft/world/storage/MapData.java
+++ b/net/minecraft/world/storage/MapData.java
@@ -6,6 +6,7 @@
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
+import java.util.UUID;
 import javax.annotation.Nullable;
 import net.minecraft.entity.item.ItemFrameEntity;
 import net.minecraft.entity.player.PlayerEntity;
@@ -27,6 +28,8 @@
 import net.minecraft.world.World;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.craftbukkit.CraftServer;
+import org.bukkit.craftbukkit.map.CraftMapView;
 
 public class MapData extends WorldSavedData {
    private static final Logger field_237240_k_ = LogManager.getLogger();
@@ -44,6 +47,12 @@
    public final Map<String, MapDecoration> field_76203_h = Maps.newLinkedHashMap();
    private final Map<String, MapFrame> field_212442_l = Maps.newHashMap();
 
+   // CraftBukkit start
+   public final CraftMapView mapView = null;
+   private CraftServer server;
+   private UUID uniqueId = null;
+   // CraftBukkit end
+
    public MapData(String p_i2140_1_) {
       super(p_i2140_1_);
    }
