--- a/net/minecraft/inventory/container/SmithingTableContainer.java
+++ b/net/minecraft/inventory/container/SmithingTableContainer.java
@@ -11,6 +11,7 @@
 import net.minecraft.item.crafting.SmithingRecipe;
 import net.minecraft.util.IWorldPosCallable;
 import net.minecraft.world.World;
+import org.bukkit.inventory.InventoryView;
 
 public class SmithingTableContainer extends AbstractRepairContainer {
    private final World field_234651_g_;
@@ -72,7 +73,12 @@
       });
    }
 
-   public boolean func_94530_a(ItemStack p_94530_1_, Slot p_94530_2_) {
+    @Override
+    public InventoryView getBukkitView() {
+        return null;
+    }
+
+    public boolean func_94530_a(ItemStack p_94530_1_, Slot p_94530_2_) {
       return p_94530_2_.field_75224_c != this.field_234642_c_ && super.func_94530_a(p_94530_1_, p_94530_2_);
    }
 }
