--- a/net/minecraft/inventory/container/SmithingTableContainer.java
+++ b/net/minecraft/inventory/container/SmithingTableContainer.java
@@ -11,12 +11,17 @@
 import net.minecraft.item.crafting.SmithingRecipe;
 import net.minecraft.util.IWorldPosCallable;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
+import org.bukkit.inventory.InventoryView;
 
 public class SmithingTableContainer extends AbstractRepairContainer {
    private final World field_234651_g_;
    @Nullable
    private SmithingRecipe field_234652_h_;
    private final List<SmithingRecipe> field_241443_i_;
+   // CraftBukkit start
+   private CraftInventoryView bukkitEntity;
+   // CraftBukkit end
 
    public SmithingTableContainer(int p_i231590_1_, PlayerInventory p_i231590_2_) {
       this(p_i231590_1_, p_i231590_2_, IWorldPosCallable.field_221489_a);
@@ -68,4 +73,16 @@
          return p_241444_1_.func_241456_a_(p_241210_1_);
       });
    }
+
+   // CraftBukkit start
+   @Override
+   public CraftInventoryView getBukkitView() {
+      if (bukkitEntity != null) {
+         return bukkitEntity;
+      }
+      org.bukkit.craftbukkit.inventory.CraftInventory inventory = new org.bukkit.craftbukkit.inventory.CraftInventorySmithing(this.field_234643_d_, this.field_234642_c_);
+      bukkitEntity = new CraftInventoryView(this.field_234645_f_.getBukkitEntity(), inventory, this);
+      return bukkitEntity;
+   }
+   // CraftBukkit end
 }
