--- a/net/minecraft/tileentity/BeaconTileEntity.java
+++ b/net/minecraft/tileentity/BeaconTileEntity.java
@@ -37,21 +37,22 @@
 import net.minecraft.world.gen.Heightmap;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.potion.PotionEffect;
 
 public class BeaconTileEntity extends TileEntity implements INamedContainerProvider, ITickableTileEntity {
    public static final Effect[][] field_146009_a = new Effect[][]{{Effects.field_76424_c, Effects.field_76422_e}, {Effects.field_76429_m, Effects.field_76430_j}, {Effects.field_76420_g}, {Effects.field_76428_l}};
    private static final Set<Effect> field_184280_f = Arrays.stream(field_146009_a).flatMap(Arrays::stream).collect(Collectors.toSet());
    private List<BeaconTileEntity.BeamSegment> field_174909_f = Lists.newArrayList();
    private List<BeaconTileEntity.BeamSegment> field_213934_g = Lists.newArrayList();
-   private int field_146012_l;
+   public int field_146012_l; // private->public CraftBukkit
    private int field_213935_i = -1;
    @Nullable
-   private Effect field_146013_m;
+   public Effect field_146013_m; // private->public CraftBukkit
    @Nullable
-   private Effect field_146010_n;
+   public Effect field_146010_n; // private->public CraftBukkit
    @Nullable
-   private ITextComponent field_146008_p;
-   private LockCode field_213936_m = LockCode.field_180162_a;
+   public ITextComponent field_146008_p; // private->public CraftBukkit
+   public LockCode field_213936_m = LockCode.field_180162_a; // private->public CraftBukkit
    private final IIntArray field_213937_n = new IIntArray() {
       public int func_221476_a(int p_221476_1_) {
          switch(p_221476_1_) {
@@ -112,8 +113,8 @@
       for(int i1 = 0; i1 < 10 && blockpos.func_177956_o() <= l; ++i1) {
          BlockState blockstate = this.field_145850_b.func_180495_p(blockpos);
          Block block = blockstate.func_177230_c();
-         if (block instanceof IBeaconBeamColorProvider) {
-            float[] afloat = ((IBeaconBeamColorProvider)block).func_196457_d().func_193349_f();
+         float[] afloat = blockstate.getBeaconColorMultiplier(this.field_145850_b, blockpos, func_174877_v());
+         if (afloat != null) {
             if (this.field_213934_g.size() <= 1) {
                beacontileentity$beamsegment = new BeaconTileEntity.BeamSegment(afloat);
                this.field_213934_g.add(beacontileentity$beamsegment);
@@ -184,7 +185,7 @@
 
          for(int k = p_213927_1_ - i; k <= p_213927_1_ + i && flag; ++k) {
             for(int l = p_213927_3_ - i; l <= p_213927_3_ + i; ++l) {
-               if (!this.field_145850_b.func_180495_p(new BlockPos(k, j, l)).func_235714_a_(BlockTags.field_232875_ap_)) {
+               if (!this.field_145850_b.func_180495_p(new BlockPos(k, j, l)).isBeaconBase(this.field_145850_b, new BlockPos(k, j, l), func_174877_v())) {
                   flag = false;
                   break;
                }
@@ -321,4 +322,19 @@
          return this.field_177265_b;
       }
    }
+
+   // CraftBukkit - start
+   public List getHumansInRange() {
+      return null;
+   }
+
+   public PotionEffect getPrimaryEffect() {
+      return null;
+   }
+
+   public PotionEffect getSecondaryEffect() {
+      return null;
+   }
+   // CraftBukkit end
+
 }
