--- a/net/minecraft/entity/monster/ZombieVillagerEntity.java
+++ b/net/minecraft/entity/monster/ZombieVillagerEntity.java
@@ -30,6 +30,7 @@
 import net.minecraft.network.datasync.DataParameter;
 import net.minecraft.network.datasync.DataSerializers;
 import net.minecraft.network.datasync.EntityDataManager;
+import net.minecraft.potion.Effect;
 import net.minecraft.potion.EffectInstance;
 import net.minecraft.potion.Effects;
 import net.minecraft.util.ActionResultType;
@@ -45,12 +46,13 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.event.entity.EntityPotionEffectEvent;
 
 public class ZombieVillagerEntity extends ZombieEntity implements IVillagerDataHolder {
-   private static final DataParameter<Boolean> field_184739_bx = EntityDataManager.func_187226_a(ZombieVillagerEntity.class, DataSerializers.field_187198_h);
+   public static final DataParameter<Boolean> field_184739_bx = EntityDataManager.func_187226_a(ZombieVillagerEntity.class, DataSerializers.field_187198_h); // private->public CraftBukkit
    private static final DataParameter<VillagerData> field_213795_c = EntityDataManager.func_187226_a(ZombieVillagerEntity.class, DataSerializers.field_218813_q);
-   private int field_82234_d;
-   private UUID field_191992_by;
+   public int field_82234_d; // private->public CraftBukkit
+   public UUID field_191992_by; // private->public CraftBukkit
    private INBT field_223728_bB;
    private CompoundNBT field_213793_bB;
    private int field_213794_bC;
@@ -157,7 +159,7 @@
       return this.func_184212_Q().func_187225_a(field_184739_bx);
    }
 
-   private void func_191991_a(@Nullable UUID p_191991_1_, int p_191991_2_) {
+   public void func_191991_a(@Nullable UUID p_191991_1_, int p_191991_2_) { // private->public CraftBukkit
       this.field_191992_by = p_191991_1_;
       this.field_82234_d = p_191991_2_;
       this.func_184212_Q().func_187227_b(field_184739_bx, true);
@@ -318,4 +320,5 @@
    public void func_213789_a(int p_213789_1_) {
       this.field_213794_bC = p_213789_1_;
    }
+
 }
