--- a/net/minecraft/entity/effect/LightningBoltEntity.java
+++ b/net/minecraft/entity/effect/LightningBoltEntity.java
@@ -27,6 +27,7 @@
    private int field_70262_b;
    public long field_70264_a;
    private int field_70263_c;
+   public boolean isEffect; // CraftBukkit
    private boolean field_184529_d;
    @Nullable
    private ServerPlayerEntity field_204810_e;
@@ -83,6 +84,7 @@
             List<Entity> list = this.field_70170_p.func_175674_a(this, new AxisAlignedBB(this.func_226277_ct_() - 3.0D, this.func_226278_cu_() - 3.0D, this.func_226281_cx_() - 3.0D, this.func_226277_ct_() + 3.0D, this.func_226278_cu_() + 6.0D + 3.0D, this.func_226281_cx_() + 3.0D), Entity::func_70089_S);
 
             for(Entity entity : list) {
+               if (!net.minecraftforge.event.ForgeEventFactory.onEntityStruckByLightning(entity, this))
                entity.func_241841_a((ServerWorld)this.field_70170_p, this);
             }
 
