--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -50,6 +50,8 @@
 import java.util.function.Function;
 import javax.annotation.Nullable;
 import javax.imageio.ImageIO;
+import jline.console.ConsoleReader;
+import joptsimple.OptionSet;
 import net.minecraft.advancements.AdvancementManager;
 import net.minecraft.advancements.FunctionManager;
 import net.minecraft.block.Block;
@@ -146,32 +148,36 @@
 import net.minecraft.world.storage.IWorldInfo;
 import net.minecraft.world.storage.PlayerData;
 import net.minecraft.world.storage.SaveFormat;
+import net.minecraft.world.storage.ServerWorldInfo;
 import net.minecraft.world.storage.WorldSavedDataCallableSave;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
 import org.apache.commons.lang3.Validate;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.Bukkit;
+import org.bukkit.command.CommandSender;
+import org.bukkit.craftbukkit.CraftServer;
 
 public abstract class MinecraftServer extends RecursiveEventLoop<TickDelayedTask> implements ISnooperInfo, ICommandSource, AutoCloseable {
    private static final Logger field_147145_h = LogManager.getLogger();
    public static final File field_152367_a = new File("usercache.json");
    public static final WorldSettings field_213219_c = new WorldSettings("Demo World", GameType.SURVIVAL, false, Difficulty.NORMAL, false, new GameRules(), DatapackCodec.field_234880_a_);
-   protected final SaveFormat.LevelSave field_71310_m;
-   protected final PlayerData field_240766_e_;
+   public final SaveFormat.LevelSave field_71310_m;// protected->public CraftBukkit
+   public final PlayerData field_240766_e_; // protected->public CraftBukkit
    private final Snooper field_71307_n = new Snooper("server", this, Util.func_211177_b());
    private final List<Runnable> field_71322_p = Lists.newArrayList();
    private final TimeTracker field_240769_m_ = new TimeTracker(Util.field_211180_a, this::func_71259_af);
    private IProfiler field_71304_b = EmptyProfiler.field_219906_a;
    private final NetworkSystem field_147144_o;
-   private final IChunkStatusListenerFactory field_213220_d;
+   public final IChunkStatusListenerFactory field_213220_d; // private->public CraftBukkit
    private final ServerStatusResponse field_147147_p = new ServerStatusResponse();
    private final Random field_147146_q = new Random();
    private final DataFixer field_184112_s;
    private String field_71320_r;
    private int field_71319_s = -1;
-   protected final DynamicRegistries.Impl field_240767_f_;
-   private final Map<RegistryKey<World>, ServerWorld> field_71305_c = Maps.newLinkedHashMap();
+   public final DynamicRegistries.Impl field_240767_f_; // protected->public CraftBukkit
+   public final Map<RegistryKey<World>, ServerWorld> field_71305_c = Maps.newLinkedHashMap(); // private->public CraftBukkit
    private PlayerList field_71318_t;
    private volatile boolean field_71317_u = true;
    private boolean field_71316_v;
@@ -201,7 +207,7 @@
    private final GameProfileRepository field_152365_W;
    private final PlayerProfileCache field_152366_X;
    private long field_147142_T;
-   private final Thread field_175590_aa;
+   public final Thread field_175590_aa; // private->public CraftBukkit
    protected long field_211151_aa = Util.func_211177_b();
    private long field_213213_ab;
    private boolean field_213214_ac;
@@ -216,16 +222,31 @@
    private final FrameTimer field_213215_ap = new FrameTimer();
    private boolean field_205745_an;
    private float field_211152_ao;
-   private final Executor field_213217_au;
+   public final Executor field_213217_au; // private->public CraftBukkit
    @Nullable
    private String field_213218_av;
-   private DataPackRegistries field_195576_ac;
+   public DataPackRegistries field_195576_ac; // private->public CraftBukkit
    private final TemplateManager field_240765_ak_;
    protected final IServerConfiguration field_240768_i_;
 
+   // CraftBukkit start
+   public DatapackCodec datapackconfiguration;
+   public CraftServer server;
+   public OptionSet options;
+   public org.bukkit.command.ConsoleCommandSender console;
+   public org.bukkit.command.RemoteConsoleCommandSender remoteConsole;
+   public ConsoleReader reader;
+   public static int currentTick = (int) (System.currentTimeMillis() / 50);
+   public java.util.Queue<Runnable> processQueue = new java.util.concurrent.ConcurrentLinkedQueue<Runnable>();
+   public int autosavePeriod;
+   public File bukkitDataPackFolder;
+   public Commands vanillaCommandDispatcher;
+   private boolean forceTicks;
+   // CraftBukkit end
+
    public static <S extends MinecraftServer> S func_240784_a_(Function<Thread, S> p_240784_0_) {
       AtomicReference<S> atomicreference = new AtomicReference<>();
-      Thread thread = new Thread(() -> {
+      Thread thread = new Thread(net.minecraftforge.fml.common.thread.SidedThreadGroups.SERVER, () -> {
          atomicreference.get().func_240802_v_();
       }, "Server thread");
       thread.setUncaughtExceptionHandler((p_240779_0_, p_240779_1_) -> {
@@ -302,79 +323,82 @@
       this.func_175584_a();
       this.field_240768_i_.func_230412_a_(this.getServerModName(), this.func_230045_q_().isPresent());
       IChunkStatusListener ichunkstatuslistener = this.field_213220_d.create(11);
-      this.func_240787_a_(ichunkstatuslistener);
+//      this.func_240787_a_(ichunkstatuslistener); // TODO: 18/08/2020 Magma
       this.func_230543_p_();
-      this.func_213186_a(ichunkstatuslistener);
+      this.loadInitialChunks(ichunkstatuslistener, null);
    }
 
    protected void func_230543_p_() {
    }
 
-   protected void func_240787_a_(IChunkStatusListener p_240787_1_) {
-      IServerWorldInfo iserverworldinfo = this.field_240768_i_.func_230407_G_();
-      DimensionGeneratorSettings dimensiongeneratorsettings = this.field_240768_i_.func_230418_z_();
-      boolean flag = dimensiongeneratorsettings.func_236227_h_();
-      long i = dimensiongeneratorsettings.func_236221_b_();
-      long j = BiomeManager.func_235200_a_(i);
-      List<ISpecialSpawner> list = ImmutableList.of(new PhantomSpawner(), new PatrolSpawner(), new CatSpawner(), new VillageSiege(), new WanderingTraderSpawner(iserverworldinfo));
-      SimpleRegistry<Dimension> simpleregistry = dimensiongeneratorsettings.func_236224_e_();
-      Dimension dimension = simpleregistry.func_230516_a_(Dimension.field_236053_b_);
-      ChunkGenerator chunkgenerator;
-      DimensionType dimensiontype;
-      if (dimension == null) {
-         dimensiontype = this.field_240767_f_.func_230520_a_().func_243576_d(DimensionType.field_235999_c_);
-         chunkgenerator = DimensionGeneratorSettings.func_242750_a(this.field_240767_f_.func_243612_b(Registry.field_239720_u_), this.field_240767_f_.func_243612_b(Registry.field_243549_ar), (new Random()).nextLong());
-      } else {
-         dimensiontype = dimension.func_236063_b_();
-         chunkgenerator = dimension.func_236064_c_();
-      }
+   // CraftBukkit start 
+   // TODO: 18/08/2020 Magma finish method
+   public void initWorld(ServerWorld worldserver, ServerWorldInfo iworlddataserver, IServerConfiguration saveData, DimensionGeneratorSettings generatorsettings) {
+//      IServerWorldInfo iserverworldinfo = this.field_240768_i_.func_230407_G_();
+//      DimensionGeneratorSettings dimensiongeneratorsettings = this.field_240768_i_.func_230418_z_();
+//      boolean flag = dimensiongeneratorsettings.func_236227_h_();
+//      long i = dimensiongeneratorsettings.func_236221_b_();
+//      long j = BiomeManager.func_235200_a_(i);
+//      List<ISpecialSpawner> list = ImmutableList.of(new PhantomSpawner(), new PatrolSpawner(), new CatSpawner(), new VillageSiege(), new WanderingTraderSpawner(iserverworldinfo));
+//      SimpleRegistry<Dimension> simpleregistry = dimensiongeneratorsettings.func_236224_e_();
+//      Dimension dimension = simpleregistry.func_230516_a_(Dimension.field_236053_b_);
+//      ChunkGenerator chunkgenerator;
+//      DimensionType dimensiontype;
+//      if (dimension == null) {
+//         dimensiontype = this.field_240767_f_.func_230520_a_().func_243576_d(DimensionType.field_235999_c_);
+//         chunkgenerator = DimensionGeneratorSettings.func_242750_a(this.field_240767_f_.func_243612_b(Registry.field_239720_u_), this.field_240767_f_.func_243612_b(Registry.field_243549_ar), (new Random()).nextLong());
+//      } else {
+//         dimensiontype = dimension.func_236063_b_();
+//         chunkgenerator = dimension.func_236064_c_();
+//      }
+//
+//      ServerWorld serverworld = new ServerWorld(this, this.backgroundExecutor, this.anvilConverterForAnvilFile, iserverworldinfo, World.field_234918_g_, dimensiontype, p_240787_1_, chunkgenerator, flag, j, list, true);
+//      this.worlds.put(World.field_234918_g_, serverworld);
+//      DimensionSavedDataManager dimensionsaveddatamanager = serverworld.getSavedData();
+//      this.func_213204_a(dimensionsaveddatamanager);
+//      this.field_229733_al_ = new CommandStorage(dimensionsaveddatamanager);
+//      WorldBorder worldborder = serverworld.getWorldBorder();
+//      worldborder.func_235926_a_(iserverworldinfo.func_230398_q_());
+//      if (!iserverworldinfo.isInitialized()) {
+//         try {
+//            func_240786_a_(serverworld, iserverworldinfo, dimensiongeneratorsettings.func_236223_d_(), flag, true);
+//            iserverworldinfo.setInitialized(true);
+//            if (flag) {
+//               this.func_240778_a_(this.field_240768_i_);
+//            }
+//         } catch (Throwable throwable1) {
+//            CrashReport crashreport = CrashReport.makeCrashReport(throwable1, "Exception initializing level");
+//
+//            try {
+//               serverworld.fillCrashReport(crashreport);
+//            } catch (Throwable throwable) {
+//            }
+//
+//            throw new ReportedException(crashreport);
+//         }
+//
+//         iserverworldinfo.setInitialized(true);
+//      }
+//
+//      this.getPlayerList().func_212504_a(serverworld);
+//      if (this.field_240768_i_.func_230404_D_() != null) {
+//         this.getCustomBossEvents().read(this.field_240768_i_.func_230404_D_());
+//      }
+//
+//      for(Entry<RegistryKey<Dimension>, Dimension> entry : simpleregistry.func_239659_c_()) {
+//         RegistryKey<Dimension> registrykey = entry.getKey();
+//         if (registrykey != Dimension.field_236053_b_) {
+//            RegistryKey<World> registrykey1 = RegistryKey.func_240903_a_(Registry.field_239699_ae_, registrykey.func_240901_a_());
+//            DimensionType dimensiontype1 = entry.getValue().func_236063_b_();
+//            ChunkGenerator chunkgenerator1 = entry.getValue().func_236064_c_();
+//            DerivedWorldInfo derivedworldinfo = new DerivedWorldInfo(this.field_240768_i_, iserverworldinfo);
+//            ServerWorld serverworld1 = new ServerWorld(this, this.backgroundExecutor, this.anvilConverterForAnvilFile, derivedworldinfo, registrykey1, dimensiontype1, p_240787_1_, chunkgenerator1, flag, j, ImmutableList.of(), false);
+//            worldborder.addListener(new IBorderListener.Impl(serverworld1.getWorldBorder()));
+//            this.worlds.put(registrykey1, serverworld1);
+//         }
+//         net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.WorldEvent.Load(worlds.get(registrykey)));
+//      }
 
-      ServerWorld serverworld = new ServerWorld(this, this.field_213217_au, this.field_71310_m, iserverworldinfo, World.field_234918_g_, dimensiontype, p_240787_1_, chunkgenerator, flag, j, list, true);
-      this.field_71305_c.put(World.field_234918_g_, serverworld);
-      DimensionSavedDataManager dimensionsaveddatamanager = serverworld.func_217481_x();
-      this.func_213204_a(dimensionsaveddatamanager);
-      this.field_229733_al_ = new CommandStorage(dimensionsaveddatamanager);
-      WorldBorder worldborder = serverworld.func_175723_af();
-      worldborder.func_235926_a_(iserverworldinfo.func_230398_q_());
-      if (!iserverworldinfo.func_76070_v()) {
-         try {
-            func_240786_a_(serverworld, iserverworldinfo, dimensiongeneratorsettings.func_236223_d_(), flag, true);
-            iserverworldinfo.func_76091_d(true);
-            if (flag) {
-               this.func_240778_a_(this.field_240768_i_);
-            }
-         } catch (Throwable throwable1) {
-            CrashReport crashreport = CrashReport.func_85055_a(throwable1, "Exception initializing level");
-
-            try {
-               serverworld.func_72914_a(crashreport);
-            } catch (Throwable throwable) {
-            }
-
-            throw new ReportedException(crashreport);
-         }
-
-         iserverworldinfo.func_76091_d(true);
-      }
-
-      this.func_184103_al().func_212504_a(serverworld);
-      if (this.field_240768_i_.func_230404_D_() != null) {
-         this.func_201300_aS().func_201381_a(this.field_240768_i_.func_230404_D_());
-      }
-
-      for(Entry<RegistryKey<Dimension>, Dimension> entry : simpleregistry.func_239659_c_()) {
-         RegistryKey<Dimension> registrykey = entry.getKey();
-         if (registrykey != Dimension.field_236053_b_) {
-            RegistryKey<World> registrykey1 = RegistryKey.func_240903_a_(Registry.field_239699_ae_, registrykey.func_240901_a_());
-            DimensionType dimensiontype1 = entry.getValue().func_236063_b_();
-            ChunkGenerator chunkgenerator1 = entry.getValue().func_236064_c_();
-            DerivedWorldInfo derivedworldinfo = new DerivedWorldInfo(this.field_240768_i_, iserverworldinfo);
-            ServerWorld serverworld1 = new ServerWorld(this, this.field_213217_au, this.field_71310_m, derivedworldinfo, registrykey1, dimensiontype1, p_240787_1_, chunkgenerator1, flag, j, ImmutableList.of(), false);
-            worldborder.func_177737_a(new IBorderListener.Impl(serverworld1.func_175723_af()));
-            this.field_71305_c.put(registrykey1, serverworld1);
-         }
-      }
-
    }
 
    private static void func_240786_a_(ServerWorld p_240786_0_, IServerWorldInfo p_240786_1_, boolean p_240786_2_, boolean p_240786_3_, boolean p_240786_4_) {
@@ -384,6 +408,7 @@
       } else if (p_240786_3_) {
          p_240786_1_.func_176143_a(BlockPos.field_177992_a.func_177984_a(), 0.0F);
       } else {
+         if (net.minecraftforge.event.ForgeEventFactory.onCreateWorldSpawn(p_240786_0_, p_240786_1_)) return;
          BiomeProvider biomeprovider = chunkgenerator.func_202090_b();
          Random random = new Random(p_240786_0_.func_72905_C());
          BlockPos blockpos = biomeprovider.func_225531_a_(0, p_240786_0_.func_181545_F(), 0, 256, (p_244265_0_) -> {
@@ -448,7 +473,8 @@
       iserverworldinfo.func_230392_a_(GameType.SPECTATOR);
    }
 
-   private void func_213186_a(IChunkStatusListener p_213186_1_) {
+   // CraftBukkit start
+   public void loadInitialChunks(IChunkStatusListener p_213186_1_, ServerWorld serverWorldl) { // private->public CraftBukkit
       ServerWorld serverworld = this.func_241755_D_();
       field_147145_h.info("Preparing start region for dimension {}", (Object)serverworld.func_234923_W_().func_240901_a_());
       BlockPos blockpos = serverworld.func_241135_u_();
@@ -538,6 +564,16 @@
       this.func_71260_j();
    }
 
+   // CraftBukkit start
+   private boolean hasStopped = false;
+   private final Object stopLock = new Object();
+   public final boolean hasStopped() {
+      synchronized (stopLock) {
+         return hasStopped;
+      }
+   }
+   // CraftBukkit end
+
    protected void func_71260_j() {
       field_147145_h.info("Stopping server");
       if (this.func_147137_ag() != null) {
@@ -563,6 +599,7 @@
       for(ServerWorld serverworld1 : this.func_212370_w()) {
          if (serverworld1 != null) {
             try {
+               net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.WorldEvent.Unload(serverworld1));
                serverworld1.close();
             } catch (IOException ioexception1) {
                field_147145_h.error("Exception closing the level", (Throwable)ioexception1);
@@ -611,6 +648,7 @@
    protected void func_240802_v_() {
       try {
          if (this.func_71197_b()) {
+            net.minecraftforge.fml.server.ServerLifecycleHooks.handleServerStarted(this);
             this.field_211151_aa = Util.func_211177_b();
             this.field_147147_p.func_151315_a(new StringTextComponent(this.field_71286_C));
             this.field_147147_p.func_151321_a(new ServerStatusResponse.Version(SharedConstants.func_215069_a().getName(), SharedConstants.func_215069_a().getProtocolVersion()));
@@ -640,7 +678,10 @@
                this.func_240795_b_(longtickdetector);
                this.field_71296_Q = true;
             }
+            net.minecraftforge.fml.server.ServerLifecycleHooks.handleServerStopping(this);
+            net.minecraftforge.fml.server.ServerLifecycleHooks.expectServerStopped(); // has to come before finalTick to avoid race conditions
          } else {
+            net.minecraftforge.fml.server.ServerLifecycleHooks.expectServerStopped(); // has to come before finalTick to avoid race conditions
             this.func_71228_a((CrashReport)null);
          }
       } catch (Throwable throwable1) {
@@ -659,6 +700,7 @@
             field_147145_h.error("We were unable to save this crash report to disk.");
          }
 
+         net.minecraftforge.fml.server.ServerLifecycleHooks.expectServerStopped(); // has to come before finalTick to avoid race conditions
          this.func_71228_a(crashreport);
       } finally {
          try {
@@ -667,6 +709,7 @@
          } catch (Throwable throwable) {
             field_147145_h.error("Exception stopping the server", throwable);
          } finally {
+            net.minecraftforge.fml.server.ServerLifecycleHooks.handleServerStopped(this);
             this.func_71240_o();
          }
 
@@ -768,6 +811,7 @@
 
    protected void func_71217_p(BooleanSupplier p_71217_1_) {
       long i = Util.func_211178_c();
+      net.minecraftforge.fml.hooks.BasicEventHooks.onPreServerTick();
       ++this.field_71315_w;
       this.func_71190_q(p_71217_1_);
       if (i - this.field_147142_T >= 5000000000L) {
@@ -782,6 +826,7 @@
 
          Collections.shuffle(Arrays.asList(agameprofile));
          this.field_147147_p.func_151318_b().func_151330_a(agameprofile);
+         this.field_147147_p.invalidateJson();
       }
 
       if (this.field_71315_w % 6000 == 0) {
@@ -809,6 +854,7 @@
       long i1 = Util.func_211178_c();
       this.field_213215_ap.func_181747_a(i1 - i);
       this.field_71304_b.func_76319_b();
+      net.minecraftforge.fml.hooks.BasicEventHooks.onPostServerTick();
    }
 
    protected void func_71190_q(BooleanSupplier p_71190_1_) {
@@ -816,7 +862,8 @@
       this.func_193030_aL().func_73660_a();
       this.field_71304_b.func_219895_b("levels");
 
-      for(ServerWorld serverworld : this.func_212370_w()) {
+      for(ServerWorld serverworld : this.getWorldArray()) {
+         long tickStart = Util.func_211178_c();
          this.field_71304_b.func_194340_a(() -> {
             return serverworld + " " + serverworld.func_234923_W_().func_240901_a_();
          });
@@ -827,6 +874,7 @@
          }
 
          this.field_71304_b.func_76320_a("tick");
+         net.minecraftforge.fml.hooks.BasicEventHooks.onPreWorldTick(serverworld);
 
          try {
             serverworld.func_72835_b(p_71190_1_);
@@ -835,9 +883,11 @@
             serverworld.func_72914_a(crashreport);
             throw new ReportedException(crashreport);
          }
+         net.minecraftforge.fml.hooks.BasicEventHooks.onPostWorldTick(serverworld);
 
          this.field_71304_b.func_76319_b();
          this.field_71304_b.func_76319_b();
+         perWorldTickTimes.computeIfAbsent(serverworld.func_234923_W_(), k -> new long[100])[this.field_71315_w % 100] = Util.func_211178_c() - tickStart;
       }
 
       this.field_71304_b.func_219895_b("connection");
@@ -912,7 +962,7 @@
    }
 
    public String getServerModName() {
-      return "vanilla";
+      return net.minecraftforge.fml.BrandingControl.getServerBranding();
    }
 
    public CrashReport func_71230_b(CrashReport p_71230_1_) {
@@ -925,6 +975,7 @@
       p_71230_1_.func_85056_g().func_189529_a("Data Packs", () -> {
          StringBuilder stringbuilder = new StringBuilder();
 
+               LogManager.shutdown(); // we're manually managing the logging shutdown on the server. Make sure we do it here at the end.
          for(ResourcePackInfo resourcepackinfo : this.field_195577_ad.func_198980_d()) {
             if (stringbuilder.length() > 0) {
                stringbuilder.append(", ");
@@ -1271,6 +1322,7 @@
          this.func_184103_al().func_193244_w();
          this.field_200258_al.func_240946_a_(this.field_195576_ac.func_240960_a_());
          this.field_240765_ak_.func_195410_a(this.field_195576_ac.func_240970_h_());
+         this.func_184103_al().func_181057_v().forEach(this.func_184103_al()::func_187243_f); //Forge: Fix newly added/modified commands not being sent to the client when commands reload.
       }, this);
       if (this.func_213162_bc()) {
          this.func_213161_c(completablefuture::isDone);
@@ -1280,10 +1332,13 @@
    }
 
    public static DatapackCodec func_240772_a_(ResourcePackList p_240772_0_, DatapackCodec p_240772_1_, boolean p_240772_2_) {
+      net.minecraftforge.fml.packs.ResourcePackLoader.loadResourcePacks(p_240772_0_, net.minecraftforge.fml.server.ServerLifecycleHooks::buildPackFinder);
       p_240772_0_.func_198983_a();
+      DatapackCodec.field_234880_a_.addModPacks(net.minecraftforge.common.ForgeHooks.getModPacks());
+      p_240772_1_.addModPacks(net.minecraftforge.common.ForgeHooks.getModPacks());
       if (p_240772_2_) {
-         p_240772_0_.func_198985_a(Collections.singleton("vanilla"));
-         return new DatapackCodec(ImmutableList.of("vanilla"), ImmutableList.of());
+         p_240772_0_.func_198985_a(net.minecraftforge.common.ForgeHooks.getModPacksWithVanilla());
+         return new DatapackCodec(net.minecraftforge.common.ForgeHooks.getModPacksWithVanilla(), ImmutableList.of());
       } else {
          Set<String> set = Sets.newLinkedHashSet();
 
@@ -1357,6 +1412,11 @@
       return true;
    }
 
+   @Override
+   public CommandSender getBukkitSender(CommandSource p0) {
+      return null;
+   }
+
    public RecipeManager func_199529_aN() {
       return this.field_195576_ac.func_240967_e_();
    }
@@ -1433,6 +1493,31 @@
 
    public abstract boolean func_213199_b(GameProfile p_213199_1_);
 
+   private Map<RegistryKey<World>, long[]> perWorldTickTimes = Maps.newIdentityHashMap();
+   @Nullable
+   public long[] getTickTime(RegistryKey<World> dim) {
+      return perWorldTickTimes.get(dim);
+   }
+
+   @Deprecated //Forge Internal use Only, You can screw up a lot of things if you mess with this map.
+   public synchronized Map<RegistryKey<World>, ServerWorld> forgeGetWorldMap() {
+      return this.field_71305_c;
+   }
+   private int worldArrayMarker = 0;
+   private int worldArrayLast = -1;
+   private ServerWorld[] worldArray;
+   @Deprecated //Forge Internal use Only, use to protect against concurrent modifications in the world tick loop.
+   public synchronized void markWorldsDirty() {
+      worldArrayMarker++;
+   }
+   private ServerWorld[] getWorldArray() {
+      if (worldArrayMarker == worldArrayLast && worldArray != null)
+         return worldArray;
+      worldArray = this.field_71305_c.values().stream().toArray(x -> new ServerWorld[x]);
+      worldArrayLast = worldArrayMarker;
+      return worldArray;
+   }
+
    public void func_223711_a(Path p_223711_1_) throws IOException {
       Path path = p_223711_1_.resolve("levels");
 
@@ -1561,7 +1646,19 @@
       return this.field_240768_i_;
    }
 
+   public DataPackRegistries getDataPackRegistries() {
+       return field_195576_ac;
+   }
+
    public DynamicRegistries func_244267_aX() {
       return this.field_240767_f_;
    }
+
+   // CraftBukkit - start
+   @Deprecated
+   public static MinecraftServer getServer() {
+      return (Bukkit.getServer() instanceof CraftServer) ? ((CraftServer) Bukkit.getServer()).getServer() : null;
+   }
+   // CraftBukkit - end
+
 }
