--- a/net/minecraft/entity/merchant/villager/AbstractVillagerEntity.java
+++ b/net/minecraft/entity/merchant/villager/AbstractVillagerEntity.java
@@ -36,8 +36,18 @@
 import net.minecraft.world.server.ServerWorld;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.inventory.CraftMerchant;
 
 public abstract class AbstractVillagerEntity extends AgeableEntity implements INPC, IMerchant {
+
+   // CraftBukkit start
+   private CraftMerchant craftMerchant;
+   @Override
+   public CraftMerchant getCraftMerchant() {
+      return (craftMerchant == null) ? craftMerchant = new CraftMerchant(this) : craftMerchant;
+   }
+   // CraftBukkit end
+
    private static final DataParameter<Integer> field_222820_bA = EntityDataManager.func_187226_a(AbstractVillagerEntity.class, DataSerializers.field_187192_b);
    @Nullable
    private PlayerEntity field_213725_bA;
@@ -165,9 +175,9 @@
    }
 
    @Nullable
-   public Entity func_241206_a_(ServerWorld p_241206_1_) {
+   public Entity changeDimension(ServerWorld p_241206_1_, net.minecraftforge.common.util.ITeleporter teleporter) {
       this.func_213750_eg();
-      return super.func_241206_a_(p_241206_1_);
+      return super.changeDimension(p_241206_1_, teleporter);
    }
 
    protected void func_213750_eg() {
