--- a/net/minecraft/entity/merchant/villager/WanderingTraderEntity.java
+++ b/net/minecraft/entity/merchant/villager/WanderingTraderEntity.java
@@ -43,6 +43,11 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.util.math.vector.Vector3d;
 import net.minecraft.world.World;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.inventory.CraftMerchant;
+import org.bukkit.craftbukkit.inventory.CraftMerchantRecipe;
+import org.bukkit.entity.AbstractVillager;
+import org.bukkit.event.entity.VillagerAcquireTradeEvent;
 
 public class WanderingTraderEntity extends AbstractVillagerEntity {
    @Nullable
@@ -120,7 +125,15 @@
          VillagerTrades.ITrade villagertrades$itrade = avillagertrades$itrade1[i];
          MerchantOffer merchantoffer = villagertrades$itrade.func_221182_a(this, this.field_70146_Z);
          if (merchantoffer != null) {
-            merchantoffers.add(merchantoffer);
+            // CraftBukkit start
+            final VillagerAcquireTradeEvent event = new VillagerAcquireTradeEvent((AbstractVillager) this.getBukkitEntity(), merchantoffer.asBukkit());
+            if (this.valid) {
+               Bukkit.getPluginManager().callEvent(event);
+            }
+            if (!event.isCancelled()) {
+               merchantoffers.add(CraftMerchantRecipe.fromBukkit(event.getRecipe()).toMinecraft());
+            }
+            // CraftBukkit end
          }
 
       }
@@ -185,6 +198,11 @@
       return SoundEvents.field_219724_nc;
    }
 
+   @Override
+   public CraftMerchant getCraftMerchant() {
+      return null;
+   }
+
    public void func_213728_s(int p_213728_1_) {
       this.field_213737_bB = p_213728_1_;
    }
