--- a/net/minecraft/entity/passive/PigEntity.java
+++ b/net/minecraft/entity/passive/PigEntity.java
@@ -47,6 +47,7 @@
 import net.minecraft.world.World;
 import net.minecraftforge.api.distmarker.Dist;
 import net.minecraftforge.api.distmarker.OnlyIn;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class PigEntity extends AnimalEntity implements IRideable, IEquipable {
    private static final DataParameter<Boolean> field_184763_bv = EntityDataManager.func_187226_a(PigEntity.class, DataSerializers.field_187198_h);
@@ -97,6 +98,21 @@
       super.func_184206_a(p_184206_1_);
    }
 
+   // CraftBukkit - start
+   public boolean getSaddled() {
+      return this.field_70180_af.func_187225_a(field_184763_bv);
+   }
+
+   public void setSaddled(boolean saddled) {
+      if (saddled) {
+         this.field_70180_af.func_187227_b(field_184763_bv, true);
+      } else {
+         this.field_70180_af.func_187227_b(field_184763_bv, false);
+      }
+
+   }
+   // CraftBukkit - end
+
    protected void func_70088_a() {
       super.func_70088_a();
       this.field_70180_af.func_187214_a(field_184763_bv, false);
@@ -214,7 +230,13 @@
          }
 
          zombifiedpiglinentity.func_110163_bv();
-         this.field_70170_p.func_217376_c(zombifiedpiglinentity);
+         // CraftBukkit start
+         if (CraftEventFactory.callPigZapEvent(this, p_70077_1_, zombifiedpiglinentity).isCancelled()) {
+            return;
+         }
+         // CraftBukkit - added a reason for spawning this creature
+         this.field_70170_p.addEntity(zombifiedpiglinentity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.LIGHTNING);
+         // CraftBukkit end
          this.func_70106_y();
       } else {
          super.func_70077_a(p_70077_1_);
