--- a/net/minecraft/inventory/Inventory.java
+++ b/net/minecraft/inventory/Inventory.java
@@ -10,12 +10,18 @@
 import net.minecraft.nbt.CompoundNBT;
 import net.minecraft.nbt.ListNBT;
 import net.minecraft.util.NonNullList;
+import org.bukkit.Location;
+import org.bukkit.craftbukkit.entity.CraftHumanEntity;
+import org.bukkit.entity.HumanEntity;
+import org.bukkit.inventory.InventoryHolder;
 
 public class Inventory implements IInventory, IRecipeHelperPopulator {
    private final int field_70481_b;
    private final NonNullList<ItemStack> field_70482_c;
    private List<IInventoryChangedListener> field_70480_d;
 
+   protected org.bukkit.inventory.InventoryHolder bukkitOwner;
+
    public Inventory(int p_i50397_1_) {
       this.field_70481_b = p_i50397_1_;
       this.field_70482_c = NonNullList.func_191197_a(p_i50397_1_, ItemStack.field_190927_a);
@@ -124,6 +130,11 @@
       this.func_70296_d();
    }
 
+   @Override
+   public int func_70297_j_() {
+      return 0;
+   }
+
    public int func_70302_i_() {
       return this.field_70481_b;
    }
@@ -151,6 +162,41 @@
       return true;
    }
 
+   @Override
+   public List<ItemStack> getContents() {
+      return null;
+   }
+
+   @Override
+   public void onOpen(CraftHumanEntity who) {
+
+   }
+
+   @Override
+   public void onClose(CraftHumanEntity who) {
+
+   }
+
+   @Override
+   public List<HumanEntity> getViewers() {
+      return null;
+   }
+
+   @Override
+   public InventoryHolder getOwner() {
+      return null;
+   }
+
+   @Override
+   public void setMaxStackSize(int size) {
+
+   }
+
+   @Override
+   public Location getLocation() {
+      return null;
+   }
+
    public void func_174888_l() {
       this.field_70482_c.clear();
       this.func_70296_d();
